src/program/config/data_format/data_format.lua

```diff
@@ -0,0 +1,135 @@
+-- Use of this source code is governed by the Apache 2.0 license; see COPYING.
+module(..., package.seeall)
+
+local lib = require("core.lib")
+local schema = require("lib.yang.schema")
+local yang_data = require("lib.yang.data")
+local usage = require("program.config.data_format.README_inc")
+
+-- Number of spaces a tab should consist of when indenting config.
+local tab_spaces = 4
+
+local function show_usage(status)
+   print(require("program.config.data_format.README_inc"))
+   main.exit(status)
+end
+
+local function parse_args(args)
+   local handlers = {}
+   handlers.h = function() show_usage(0) end
+   args = lib.dogetopt(args, handlers, "h", {help="h"})
+   if #args <= 0 then show_usage(1) end
+   return unpack(args)
+end
+
+function run(args)
+   function print_level(level, ...)
+      io.write(string.rep(" ", level * tab_spaces))
+      print(...)
+   end
+   function union_type(union)
+      local rtn
+      for _, t in pairs(union.argument_type.union) do
+         if rtn then
+            rtn = rtn .. " | " .. t.argument_string
+         else
+            rtn = t.argument_string
+         end
+      end
+      return rtn
+   end
+   function collate_options(name, node)
+      local options = {}
+      if node.keys ~= nil then
+	 for _, k in pairs(node.keys) do
+	    if name == k then options.key = true end
+	 end
+      end
+      if node.argument_type then
```

Comment:
I think it's not the right thing to collate options on all productions.  There are only 4 production kinds: table, array, struct, and scalar.  The set of properties of interest on these productions is fixed.  E.g. argument_type is only a thing on a scalar production.  keys is only a property on tables, mandatory is only on scalars, etc.  Can you push these things down into the handlers for the various kinds please?